v3/lints/cabf_br/lint_ca_invalid_eku.go

```diff
@@ -0,0 +1,81 @@
+/*
+ * ZLint Copyright 2024 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/*
+ * Contributed by Adriano Santoni <asantoni64@gmail.com>
+ */
+
+package cabf_br
+
+import (
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+
+	"fmt"
+)
+
+func init() {
+	lint.RegisterCertificateLint(&lint.CertificateLint{
+		LintMetadata: lint.LintMetadata{
+			Name:          "e_ca_invalid_eku",
+			Description:   "Checks that SubCA certificates do not contain forbidden values in their EKU extension.",
+			Citation:      "CABF BRs §7.1.2",
+			Source:        lint.CABFBaselineRequirements,
+			EffectiveDate: util.CABFBRs_1_7_1_Date,
+		},
+		Lint: NewCaInvalidEKU,
+	})
+}
+
+type caInvalidEKU struct{}
+
+func NewCaInvalidEKU() lint.LintInterface {
+	return &caInvalidEKU{}
+}
+
+func (l *caInvalidEKU) CheckApplies(c *x509.Certificate) bool {
+	// This lint applies to any SubCA certificate to which the CABF BRs are applicable and which contains
+	// the EKU extension. Given that the lint source is lint.CABFBaselineRequirements, if we arrive here
+	// it's been already checked that the certificate falls within the purview of the CABF BRs.
+	return util.IsSubCA(c) && len(c.ExtKeyUsage) != 0
+}
+
+func (l *caInvalidEKU) Execute(c *x509.Certificate) *lint.LintResult {
+
+	// If the EKU contains anyExtendedKeyUsage, it's probably a cross-certicate
+	// In this case, the EKU must not contain any other value
```

Comment:
I can see why you conditionally let `anyExtUsage` pass, however this does run contradictory to the clause...

> anyExtendedKeyUsage [RFC5280] MUST NOT be present. 

If the `the EKU must not contain any other value` is coming from CABF then that's problematic. However, if this come from a different doc I think that we should prefer the explicit failure spelled out by CABF here.

Reply:
Yes, I understand the doubts you may have regarding the reference version of the CABF BR for this lint. I myself was uncertain about it. The fact is, the ban on having certain other keyPurposes in the EKU of a TLS SubCA certificate was introduced in version 1.7.1 (section 7.1.2.2, subsection 'g'). However, the list of expressly prohibited keyPurposes has been extended in BRs 2.0.0 (section 7.1.2.10.6) by adding `ocspSigning`, hence the list of banned keyPurposes you see in my lint. So the requirement goes back "mostly" to BR 1.7.1, so to speak. In fact you can find EKU related incidents on https://bugzilla.mozilla.org in SubCA certificates that mention BR v1.7.1.

And as regards the first check you find in my lint, where we don't know with mathematical certainty but we can reasonably guess that we are analyzing a cross-certificate, there I refer to section 7.1.2.2.4 of the BR since v2.0.0 (Cross‑ Certified Subordinate CA Extended Key Usage - Unrestricted).

So how do you see it?